src/extractor/guidance/intersection_generator.cpp

```diff
@@ -303,6 +421,110 @@ Intersection IntersectionGenerator::mergeSegregatedRoads(Intersection intersecti
     return intersection;
 }
 
+// OSM can have some very steep angles for joining roads. Considering the following intersection:
+//        x
+//        |__________c
+//       /
+// a ---d
+//       \ __________b
+//
+// with c->d as a oneway
+// and d->b as a oneway, the turn von x->d is actually a turn from x->a. So when looking at the
+// intersection coming from x, we want to interpret the situation as
+//           x
+// a __ d __ |__________c
+//      |
+//      |_______________b
+//
+// Where we see the turn to `d` as a right turn, rather than going straight.
+// We do this by adjusting the local turn angle at `x` to turn onto `d` to be reflective of this
+// situation.
+Intersection IntersectionGenerator::AdjustForJoiningRoads(const NodeID node_at_intersection,
+                                                          Intersection intersection) const
+{
+    // nothing to do for dead ends
+    if (intersection.size() <= 1)
+        return intersection;
+
+    // never adjust u-turns
+    for( std::size_t index = 1; index < intersection.size(); ++index )
+    {
+        auto &road = intersection[index];
+        // to find out about the above situation, we need to look at the next intersection (at d in
+        // the example). If the initial road can be merged to the left/right, we are about to adjust
+        // the angle.
+        const auto next_intersection_along_road =
+            GetConnectedRoads(node_at_intersection, road.turn.eid);
+        if (next_intersection_along_road.size() <= 1)
+            continue;
+
+        const auto adjustAngle = [](double angle, double offset) {
```

Comment:
This could be a general helper function for angles. `normalizeAngle(double angle)` that normalizes to `[0, 359]`. `normalizeAngle(angle + offset)` would do the same as the function here.